Differential expression analysis
========================================================



John Blischak

Genomics and Sytems Biology, University of Chicago

Date: 2021-08-24

In this tutorial, we will perform a basic differential expression analysis with RNA sequencing data using R/Bioconductor.
Before class, please download the data set and install the software as explained in the following section.

## Download data and install software

For this tutorial, we will use the data set generated by the Sequencing Quality Control ([SEQC][]) project.
Please download the data by running the line of R code below.
If this fails, you can try changing the `method` parameter, e.g. `method = "auto"`, `method = "wget"` or `method = "curl"`.
And as a last resort, you can download it by clicking on this [link][geo].

[SEQC]: http://www.fda.gov/ScienceResearch/BioinformaticsTools/MicroarrayQualityControlProject/
[geo]: https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE49712&format=file&file=GSE49712_HTSeq.txt.gz


```r
fname <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE49712&format=file&file=GSE49712_HTSeq.txt.gz"
download.file(fname, destfile = "GSE49712_HTSeq.txt.gz", method = "curl")
```

Next, we need to download and install the two [Bioconductor][bioc] packages we will use to perform the analysis.

[bioc]: http://www.bioconductor.org


```r
source("http://bioconductor.org/biocLite.R")
biocLite(c("edgeR", "DESeq2"))
```










































