# Practical 3. Integration of scRNA-seq data from different species using CCA

We will apply Canonical Correlation Analysis (CCA) in pancreas data shown in @pmid29608179. This is done to integrate single-cell RNA-seq datasets from human and mouse. 

## Load count matrices, normalisation and CCA
 

```{r, eval=TRUE, message=FALSE}
#results="hide" above
human.data <- read.table("data/pancreas_human.expressionMatrix.txt", sep = "\t")
mouse.data <- read.table("data/pancreas_mouse.expressionMatrix.txt", sep = "\t")

# setup Seurat objects
# since both count matrices have already filtered cells, we do no 
# additional filtering here

library(Seurat)

human <- CreateSeuratObject(raw.data = human.data)
human <- NormalizeData(human)
human <- ScaleData(human)
human <- FindVariableGenes(human, do.plot = F)

mouse <- CreateSeuratObject(raw.data = mouse.data)
mouse <- NormalizeData(mouse)
mouse <- ScaleData(mouse)
mouse <- FindVariableGenes(mouse, do.plot = F)

hvg.human <- rownames(head(human@hvg.info, 1000))
hvg.mouse <- rownames(head(mouse@hvg.info, 1000))
hvg.union <- union(hvg.human, hvg.mouse)

# now we scale the data, adding in the individual (stored in orig.ident) 
# as a regression covariate
mouse <- ScaleData(mouse, vars.to.regress = "orig.ident", genes.use = hvg.union)
human <- ScaleData(human, vars.to.regress = "orig.ident", genes.use = hvg.union)

mouse@meta.data[, "species"] <- "mouse"
human@meta.data[, "species"] <- "human"

# now we run CCA
pancreas.combined <- RunCCA(mouse, human, genes.use = hvg.union)
PrintDim(pancreas.combined, reduction.type = "cca")
pancreas.combined@meta.data[mouse@cell.names, "species"] <- "mouse"
pancreas.combined <- CalcVarExpRatio(pancreas.combined, reduction.type = "pca", 
                                     grouping.var = "species", dims.use = 1:20)
pancreas.combined.all <- pancreas.combined
# Visualize the aligned datasets (before alignment)
pancreas.combined <-RunPCA(pancreas.combined , pc.genes = NULL, pcs.compute = 20, 
                           use.imputed = FALSE, rev.pca = FALSE, 
                           weight.by.var = TRUE, do.print = TRUE, pcs.print = 1:5, 
                           genes.print = 30, reduction.name = "pca", 
                           reduction.key = "PC", assay.type = "RNA", seed.use = 42)
pancreas.combined <- RunTSNE(pancreas.combined, reduction.use = "pca", 
                             dims.use = 1:20, do.fast = T)
TSNEPlot(pancreas.combined, group.by = "species")
#-----

pancreas.combined <- SubsetData(pancreas.combined, subset.name = "var.ratio.pca", 
                                accept.low = 0.5)
pancreas.combined <- AlignSubspace(pancreas.combined, reduction.type = "cca", 
                                   grouping.var = "species", dims.align = 1:20) #1:20
pancreas.combined <- RunTSNE(pancreas.combined, reduction.use = "cca.aligned", 
                             dims.use = 1:20, do.fast = T)

# Visualize the aligned datasets
TSNEPlot(pancreas.combined, group.by = "species")
```


```{r, eval=TRUE, message=FALSE, results='hide', echo=TRUE}
# Perform joint clustering
pancreas.combined <- FindClusters(pancreas.combined, reduction.type = "cca.aligned", 
                                  dims.use = 1:20, save.SNN = T)
```


```{r, eval=TRUE, message=FALSE}
# Visualize the clusters
TSNEPlot(pancreas.combined)
# Finally, we can also plot the UMAP
pancreas.combined <- RunUMAP(pancreas.combined, reduction.use = "cca", dims.use = 1:20)
DimPlot(pancreas.combined , reduction.use = "umap", plot.title='UMAP (20 canonical
        components)', pt.size=0.3,  do.label=TRUE, do.return = TRUE)

```

 
 **Exercise 3.1: Pancreatic $\beta$ cells are those producing insulin. Can you indentify them in the dataset? Tip: Use gene marker, e.g. PDX1.**
 
```{r, eval=TRUE, message=FALSE}
 
FeaturePlot(object = pancreas.combined, reduction.use='umap',features.plot = c("PDX1","MAFA"), min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)

```

We can now compare with the Beta cells in the paper:

```{r, eval=TRUE, message=FALSE}
# The exact tSNE and clusters can vary slightly, below we load in the output from the manuscript
pancreas.metadata <- read.table("data/Supplementary_Table_PancreasCellData.tsv", sep = "\t", skip = 2, header = T, row.names = 1)
pancreas.combined <- AddMetaData(pancreas.combined, metadata = pancreas.metadata)
TSNEPlot(pancreas.combined, group.by = "Cluster_ID", do.label=TRUE )
DimPlot(pancreas.combined ,group.by = "Cluster_ID", reduction.use = "umap", plot.title='UMAP', pt.size=0.3,  do.label=TRUE, do.return = TRUE)


```

 
**Exercise 3.2: Can you identify gene markers that are unique in the cluster of Pancreatic $\beta$ cells?**


**Exercise 3.3: Can you identify genes that are differentially expressed between Pancreatic $\beta$ cells and pancreatic $\alpha$ cells?**

## sessionInfo()

```{r echo=FALSE}
sessionInfo()
```




